drivers/ieee802154/ieee802154_alif.c

```diff
@@ -284,22 +308,74 @@ static enum alif_mac154_status_code alif_transmit_csma(const struct device *dev,
 	return ret;
 }
 
-static int alif_tx(const struct device *dev, enum ieee802154_tx_mode mode, struct net_pkt *pkt,
-		   struct net_buf *frag)
+static uint32_t alif_cls_ie_element_update(uint8_t *header_ptr, uint16_t length,
+					   uint16_t csl_period)
+{
+	struct mac_header_IE_s csl_ie;
+	struct alif_mac154_csl_phase csl_phase;
+	uint16_t phase, period;
+	uint32_t tx_timestamp;
+
+	csl_ie.id = MAC_HEADER_CSL_IE_ID;
+	if (!alif_mac154_ie_header_element_get(header_ptr, length, &csl_ie)) {
+		return 0;
+	}
+
+	period = csl_period;
+
+	/* Update CSL phase  */
+	if (alif_mac154_csl_phase_get(&csl_phase) != ALIF_MAC154_STATUS_OK) {
+		return 0;
+	}
+
+	tx_timestamp = (uint32_t)csl_phase.timestamp;
+	phase = csl_phase.csl_phase;
+	LOG_DBG("CSL Period %" PRIu16 " phase %" PRIu16 " ts %" PRIu16 "", period, phase,
+		tx_timestamp);
+
+	/* Calculate phase from 15ms since now */
+	tx_timestamp += (CSL_PHASE_DELAY_US + CSL_PHASE_TICK_US);
+	if (phase > CSL_PERIOD_DELAY) {
+		phase -= CSL_PERIOD_DELAY;
+	} else {
+		uint16_t delta_phase = CSL_PERIOD_DELAY - phase;
+
+		if (delta_phase == 0) {
+			delta_phase += 1;
+			tx_timestamp += CSL_PHASE_TICK_US;
+		}
+		phase = (period - delta_phase);
+	}
+	LOG_DBG("CSL phase %u tx time %u", phase, tx_timestamp);
+	sys_put_le16(phase, csl_ie.content_ptr);
+	sys_put_le16(period, csl_ie.content_ptr + 2);
+	return tx_timestamp;
+}
+
+static uint8_t mac_encode_msg[IEEE802154_MAX_PHY_PACKET_SIZE];
```

Comment:
Place this inside the function as it's not used anywhere else.

Reply:
I move this to 15.4 device structure